client/src/shutdown/components/ShutdownButton.tsx

```diff
@@ -0,0 +1,68 @@
+import * as React from "react";
+import { connect } from "react-redux";
+import { Button, Header, Icon, Modal } from "semantic-ui-react";
+import * as channelActions from "../../channel/actions";
+import { RootReducer } from "../../store";
+import { handleSubmit } from "../api";
+
+const mapDispatchToProps = {
+    shutdownAction: channelActions.Actions.shutdown,
+};
+
+const mapStateToProps = (state: RootReducer) => {
+    return {
+        channel: state.channelStatus.status,
+    };
+};
+
+type MergedProps = DispatchProps<typeof mapDispatchToProps> & ReturnType<typeof mapStateToProps>;
+
+class ShutdownButton extends React.Component<MergedProps> {
+    public state = {
+        modal: false,
+        shutdown: false,
+    };
+
+    public modalOpen = () => {
+        this.setState({ modal: true });
+    };
+
+    public modalClose = () => {
+        this.setState({ modal: false });
+    };
+
+    public handleShutdown = () => {
+        const timestamp = Date.now();
+        this.setState({ shutdown: true });
+        this.props.shutdownAction(timestamp)
+        handleSubmit();
+    };
```

Comment:
I gave it a spin - one thing I noticed was that the "currently shutting down" state was never displayed, even if the shutdown took a while. I think there are two things to fix: 1) in this function, the `shutdownAction` should only be dispatched once the answer from the HTTP API is received, and 2) only send the `SERVER_SHUTDOWN` message once the executor has finished closing. You can test that this works by adding an artificial `await asyncio.sleep(some_seconds)` to your `ShutdownHandler`, before sending the response.

For 1), a hint: as the `handleShuwdown`method here is not defined as `async`, you will have to use the Promise/callback API here, so something like this:

```typescript
handleSubmit().then((response) => { this.props.shutdownAction(...); });
```

(NB: `handleSubmit` could be named better - maybe `doShutdown` or similar? Something that indicates that this shuts down the server - `handleSubmit` sounds more like something that handles a form submission to me)

Reply:
Btw. the `CHANNEL_SHUTDOWN` action is dispatched two times - once here, and once from the saga - is that intentional? Maybe it should only be dispatched in the saga? Or only here?

Reply:
> is that intentional? Maybe it should only be dispatched in the saga? Or only here?

Yes, It's only temporary. Will be removed from saga after some other tweaks.